CDTDatastoreTests/CDTIamSessionCookieInterceptorTests.m

```diff
@@ -0,0 +1,593 @@
+//
+
+#import <Foundation/Foundation.h>
+
+
+#import <XCTest/XCTest.h>
+#import "CloudantSyncTests.h"
+#import "CDTIamSessionCookieInterceptor.h"
+#import "CDTURLSession.h"
+#import <OHHTTPStubs/OHHTTPStubs.h>
+#import <OHHTTPStubs/OHHTTPStubsResponse+JSON.h>
+#import <OHHTTPStubs/NSURLRequest+HTTPBodyTesting.h>
+#import "CDTLogging.h"
+
+// expose properties so we can look at them
+@interface CDTIAMSessionCookieInterceptor ()
+
+@property (nonatomic) BOOL shouldMakeSessionRequest;
+@property (nullable, strong, nonatomic) NSString *cookie;
+@property (nonnull, nonatomic, strong) NSURLSession *urlSession;
+
+@property NSURL *iamTokenUrl;
+
+
+@end
+
+static const NSString *testCookieHeaderValue =
+@"AuthSession=cm9vdDo1MEJCRkYwMjq0LO0ylOIwShrgt8y-UkhI-c6BGw";
+
+// helper to sequence a number of stubbed responses
+
+@interface OHHTTPStubsHelper : NSObject
+@property NSMutableArray<OHHTTPStubsResponseBlock> *responses;
+@property int currentResponse;
+
+- (id) init;
+- (void) addResponse:(OHHTTPStubsResponseBlock)responseBlock;
+- (void) doStubsForHost:(NSString*)host;
+@end
+
+@implementation OHHTTPStubsHelper
+
+- (id) init
+{
+    if (self = [super init]) {
+        _currentResponse = 0;
+        _responses = [NSMutableArray array];
+    }
+    return self;
+}
+
+- (void) addResponse:(OHHTTPStubsResponseBlock)responseBlock
+{
+    [_responses addObject:responseBlock];
+}
+- (void) doStubsForHost:(NSString*)host
+{
+    [OHHTTPStubs stubRequestsPassingTest:^BOOL(NSURLRequest *__nonnull request) {
+        return [[request.URL host] isEqualToString:host];
+    } withStubResponse:^OHHTTPStubsResponse *__nonnull(NSURLRequest *__nonnull request) {
+        return [_responses objectAtIndex:_currentResponse++](request);
+    }];
+}
+
+@end
+
+@interface CDTIAMSessionCookieInterceptorTests : CloudantSyncTests
+
+@end
+
+@implementation CDTIAMSessionCookieInterceptorTests
+
+- (void)setUp {
+    setenv("CDT_TEST_ENABLE_OHHTTPSTUBS", "1", true);
+    CDTChangeLogLevel(CDTTD_REMOTE_REQUEST_CONTEXT, DDLogLevelDebug);
+    CDTChangeLogLevel(CDTREPLICATION_LOG_CONTEXT, DDLogLevelDebug);
+    [DDLog addLogger:[DDTTYLogger sharedInstance]];
+}
+
+- (void)tearDown {
+    unsetenv("CDT_TEST_ENABLE_OHHTTPSTUBS");
+    [OHHTTPStubs removeAllStubs];
+}
+
+/**
+ * Test normal IAM token and IAM session request path by calling interceptRequestInContext directly
+ */
+- (void)testiamTokenAndCookieSuccessful
+{
+    
+    [OHHTTPStubs stubRequestsPassingTest:^BOOL(NSURLRequest *__nonnull request) {
+        return [[request.URL host] isEqualToString:@"iam.bluemix.net"];
+    }
+                        withStubResponse:^OHHTTPStubsResponse *__nonnull(NSURLRequest *__nonnull request) {
+                            if ([request.HTTPMethod isEqualToString:@"POST"]) {
+                                return [OHHTTPStubsResponse
+                                        responseWithJSONObject:@{@"access_token" : @"IamToken1"}
```

Comment:
Suggest some constants for these token values, and making them longer, perhaps replacing with the Java mock content or similar.